src/parquet/encodings/encodings.h

```diff
@@ -67,6 +67,38 @@ class Decoder {
   int num_values_;
 };
 
+
+// Base class for value encoders. Since encoders may or not have state (e.g.,
+// dictionary encoding) we use a class instance to maintain any state.
+//
+// TODO(wesm): Encode interface API is temporary
+template <int TYPE>
+class Encoder {
+ public:
+  typedef typename type_traits<TYPE>::value_type T;
+
+  virtual ~Encoder() {}
+
+  // Subclasses should override the ones they support
+  //
+  // @returns: the number of bytes written to dst
+  virtual size_t Encode(const T* src, int num_values, uint8_t* dst) {
```

Comment:
Are you sure you want dst to be a raw array instead of something like output stream? We can revisit later when the encoding side is built out more but this might be hard to deal with. Haven't put too much thought but maybe something blocky:
i.e.
byte* buffer = output->Reserve(big block for a batch).